src/PhpSpreadsheet/DefinedName.php

```diff
@@ -0,0 +1,261 @@
+        // If local only, then the scope will be set to worksheet unless a scope is explicitly set
+        $this->scope = ($localOnly === true) ? (($scope === null) ? $worksheet : $scope) : null;
+        // If the range string contains characters that aren't associated with the range definition (A-Z,1-9
+        //      for cell references, and $, or the range operators (colon comma or space), quotes and ! for
+        //      worksheet names
+        //  then this is treated as a named formula, and not a named range
+        $this->isFormula = self::testIfFormula($this->value);
+    }
+
+    /**
+     * Create a new defined name, either a range or a formula.
+     */
+    public static function createInstance(
+        string $name,
+        ?Worksheet $worksheet = null,
+        ?string $value = null,
+        bool $localOnly = false,
+        ?Worksheet $scope = null
+    ): self {
+        $value = (string) $value;
+        $isFormula = self::testIfFormula($value);
+        if ($isFormula) {
+            return new NamedFormula($name, $worksheet, $value, $localOnly, $scope);
+        }
+
+        return new NamedRange($name, $worksheet, $value, $localOnly, $scope);
+    }
+
+    public static function testIfFormula(string $value): bool
+    {
+        if (substr($value, 0, 1) === '=') {
+            $value = substr($value, 1);
+        }
+
+        if (is_numeric($value)) {
+            return true;
+        }
+
+        $segMatcher = false;
+        foreach (explode("'", $value) as $subVal) {
+            //    Only test in alternate array entries (the non-quoted blocks)
+            if (($segMatcher = !$segMatcher) &&
+                (preg_match('/' . self::REGEXP_IDENTIFY_FORMULA . '/miu', $subVal))) {
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    /**
+     * Get name.
+     */
+    public function getName(): string
+    {
+        return $this->name;
+    }
+
+    /**
+     * Set name.
+     */
+    public function setName(string $name): self
+    {
+        if (!empty($name)) {
+            // Old title
+            $oldTitle = $this->name;
+
+            // Re-attach
+            if ($this->worksheet !== null) {
+                $this->worksheet->getParent()->removeNamedRange($this->name, $this->worksheet);
+            }
+            $this->name = $name;
+
+            if ($this->worksheet !== null) {
+                $this->worksheet->getParent()->addNamedRange($this);
+            }
+
+            // New title
+            $newTitle = $this->name;
+            ReferenceHelper::getInstance()->updateNamedFormulas($this->worksheet->getParent(), $oldTitle, $newTitle);
+        }
+
+        return $this;
+    }
+
+    /**
+     * Get worksheet.
+     */
+    public function getWorksheet(): ?Worksheet
+    {
+        return $this->worksheet;
+    }
+
+    /**
+     * Set worksheet.
+     */
+    public function setWorksheet(?Worksheet $value = null): self
```

Comment:
I think a setter with a default value usually does not make sense. I'd rather explicitly write that the new value is `null`.

```suggestion
    public function setWorksheet(?Worksheet $value): self
```